pkg/collector/corechecks/gpu/gpu.go

```diff
@@ -0,0 +1,206 @@
+func (m *CheckConfig) Parse(data []byte) error {
+	return yaml.Unmarshal(data, m)
+}
+
+// Cancel cancels the check
+func (m *Check) Cancel() {
+	ret := nvml.Shutdown()
+	if ret != nvml.SUCCESS {
+		log.Warnf("Failed to shutdown NVML: %v", nvml.ErrorString(ret))
+	}
+}
+
+// Configure parses the check configuration and init the check
+func (m *Check) Configure(senderManager sender.SenderManager, _ uint64, config, initConfig integration.Data, source string) error {
+	if err := m.CommonConfigure(senderManager, initConfig, config, source); err != nil {
+		return err
+	}
+	if err := m.config.Parse(config); err != nil {
+		return fmt.Errorf("ebpf check config: %s", err)
+	}
+
+	return nil
+}
+
+func (m *Check) ensureInitialized() error {
+	var err error
+
+	if m.sysProbeUtil == nil {
+		m.sysProbeUtil, err = processnet.GetRemoteSystemProbeUtil(
+			pkgconfigsetup.SystemProbe().GetString("system_probe_config.sysprobe_socket"),
+		)
+		if err != nil {
+			return fmt.Errorf("sysprobe connection: %s", err)
+		}
+	}
+
+	if m.timeResolver == nil {
+		m.timeResolver, err = sectime.NewResolver()
+		if err != nil {
+			return fmt.Errorf("cannot create time resolver: %s", err)
+		}
+	}
+	return nil
+}
+
+// ensureProcessor ensures that there is a stats processor for the given key
+func (m *Check) ensureProcessor(key *model.StreamKey, snd sender.Sender, gpuThreads int, checkDuration time.Duration) {
+	if _, ok := m.statProcessors[key.Pid]; !ok {
+		m.statProcessors[key.Pid] = &StatsProcessor{
+			key: key,
+		}
+	}
+
+	m.statProcessors[key.Pid].totalThreadSecondsUsed = 0
+	m.statProcessors[key.Pid].sender = snd
+	m.statProcessors[key.Pid].gpuMaxThreads = gpuThreads
+	m.statProcessors[key.Pid].measuredInterval = checkDuration
+	m.statProcessors[key.Pid].timeResolver = m.timeResolver
+	m.statProcessors[key.Pid].lastCheck = m.lastCheckTime
+}
+
+// Run executes the check
+func (m *Check) Run() error {
+	if err := m.ensureInitialized(); err != nil {
+		return err
+	}
+
+	gpuDevices, err := getGPUDevices()
+	if err != nil {
+		return fmt.Errorf("get GPU devices: %s", err)
+	}
+
+	if len(gpuDevices) == 0 {
+		return fmt.Errorf("no GPU devices found")
+	}
+
+	data, err := m.sysProbeUtil.GetCheck(sysconfig.GPUMonitoringModule)
+	if err != nil {
+		return fmt.Errorf("get gpu check: %s", err)
+	}
+	now := time.Now()
+
+	var checkDuration time.Duration
+	// mark the check duration as close to the actual check as possible
+	if !m.lastCheckTime.IsZero() {
+		checkDuration = now.Sub(m.lastCheckTime)
+	}
+
+	snd, err := m.GetSender()
+	if err != nil {
+		return fmt.Errorf("get metric sender: %s", err)
+	}
+
+	stats, ok := data.(model.GPUStats)
+	if !ok {
+		return log.Errorf("ebpf check raw data has incorrect type: %T", stats)
+	}
+
+	// TODO: Multiple GPUs are not supported yet
+	gpuThreads, err := gpuDevices[0].GetMaxThreads()
+	if err != nil {
+		return fmt.Errorf("get GPU device threads: %s", err)
+	}
+
+	// Commit the metrics even in case of an error
+	defer snd.Commit()
```

Comment:
should this be called right after successful sender retrieval (`m.GetSender`)

Reply:
the comment above is confusing then, it suggest to have it immediately after retrieving the Sender